server/src/main/java/org/opensearch/cluster/routing/WeightedRoundRobin.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.routing;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+public class WeightedRoundRobin<T> implements Iterator<T>, Iterable<T> {
+
+    private List<WeightedRoundRobin.Entity<T>> entities;
+    private int turn;
+    private int lastSelectedEntity;
+    private double currentWeight = 0;
+
+    public WeightedRoundRobin(List<WeightedRoundRobin.Entity<T>> entities) {
+        this.entities = entities;
+        this.turn = 0;
+        this.lastSelectedEntity = -1;
+    }
+
+    /* (non-Javadoc)
+     * @see java.util.Iterator#hasNext()
+     */
+    @Override
+    public boolean hasNext() {
+        return entities.size() > 0;
+    }
+
+    /* (non-Javadoc)
+     * @see java.util.Iterator#next()
+     */
+    @Override
+    public T next() {
+        Entity<T> entity = entities.get(turn++);
+        return entity.getTarget();
+    }
+
+    /* (non-Javadoc)
+     * @see java.lang.Iterable#iterator()
+     */
+    @Override
+    public Iterator<T> iterator() {
+        return this;
+    }
+
+    /**
+     * *
+     * @return list of entities that is ordered using weighted round-robin scheduling
+     * http://kb.linuxvirtualserver.org/wiki/Weighted_Round-Robin_Scheduling
+     */
+    public List<WeightedRoundRobin.Entity<T>> orderEntities() {
```

Comment:
Are you planning to add tests specifically for this method since it is a core?

Reply:
Makes sense. Will do that.